odoo/addons/base/models/res_users.py

```diff
@@ -998,19 +998,24 @@ def create(self, vals_list):
         return super(UsersImplied, self).create(vals_list)
 
     def write(self, values):
+        if not values.get('groups_id'):
+            return super(UsersImplied, self).write(values)
         users_before = self.filtered(lambda u: u.has_group('base.group_user'))
         res = super(UsersImplied, self).write(values)
-        if values.get('groups_id'):
-            # add implied groups for all users
-            for user in self:
-                if not user.has_group('base.group_user') and user in users_before:
-                    # if we demoted a user, we strip him of all its previous privileges
-                    # (but we should not do it if we are simply adding a technical group to a portal user)
-                    vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
-                    super(UsersImplied, user).write(vals)
-                gs = set(concat(g.trans_implied_ids for g in user.groups_id))
-                vals = {'groups_id': [(4, g.id) for g in gs]}
-                super(UsersImplied, user).write(vals)
+        users_to_demote = users_before.filtered(
+            lambda u: not u.has_group('base.group_user')
+        )
+        if users_to_demote:
+            vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
```

Comment:
Please add the comment line
```suggestion
            # demoted users are restricted to the assigned groups only
            vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
```